### ISSUE TITLE
Position variable in OutParamArgument is alway zero when using named parameters.

### ISSUE THREAD

**Issue by jcshaw** [jdbi/jdbi#404]
Hello,

Using @SqlCall to call a Stored Procedure and using (Call) .registerOutParameter(name, type), the position variable is zero for any named parameter; when calling the map function on OutParamArgument getting a java.sql.SQLException: Invalid column index.

Thanks.

C.

**Comment by qualidafial** [comment]
What version of jdbi are you using. Have you tried upgrading to the latest
2.x release?

**Comment by jcshaw** [comment]
Thanks for replying, I'm using 2.73 from Maven Central, is there another version I should be using?

**Comment by qualidafial** [comment]
That's the latest--just making sure.

Can you provide a code snippet or unit test that reproduces the problem?

**Comment by jcshaw** [comment]
Sorry for the delay..

I'm fairly new with JDBI and had trouble getting the creation of the stored proc to work, I am sure you can fix it up or just insert the stored proc into your test db.

I used  Michael Seele [annotation](https://groups.google.com/forum/#!topic/jdbi/yShp_-7Q7wM) which I think is an awesome way to get out parameters and hope it will be part of a future release.

Thanks for any help.

```
package com.jdbitest;

import org.skife.jdbi.v2.*;
import org.skife.jdbi.v2.sqlobject.*;

import java.lang.annotation.*;
import java.lang.reflect.Method;
import java.sql.SQLException;
import java.sql.Types;

public class JdbiTest {
    public static void main(String[] args) {
        new JdbiTest();
    }

    public JdbiTest(){
        DBI dbi = new DBI("jdbc:oracle:thin:system/oracle@//<ip>:<port>/<db>");
        Integer integerOut = new Integer(0);

        MyDao myDao = dbi.onDemand(MyDao.class);

        //myDao.insertStoredProc();
        OutParameters outParameters = myDao.callStoredProc(integerOut);

        System.out.println("OutParameter: " + outParameters.getInt("integerOut"));
    }

    public interface MyDao{
        @SqlCall("BEGIN procOneOUTParameter(:outParam1); END;")
        @OutParameter(name="outParam1", sqlType = Types.INTEGER)
        OutParameters callStoredProc(@Bind("outParam1") Integer outParam1);

        @SqlUpdate("CREATE OR REPLACE PROCEDURE procOneOUTParameter(outParam1 OUT NUMBER) IS BEGIN outParam1 := 100; END")
        void insertStoredProc();
    }

    @SqlStatementCustomizingAnnotation(OutParameter.Factory.class)
    @Target(ElementType.METHOD)
    @Retention(RetentionPolicy.RUNTIME)
    public @interface OutParameter {
        String name();
        int sqlType();
        static class Factory implements SqlStatementCustomizerFactory {
            @Override
            public SqlStatementCustomizer createForType(Annotation annotation, @SuppressWarnings("rawtypes") Class sqlObjectType) {
                throw new UnsupportedOperationException("Not allowed on Type"); //$NON-NLS-1$
            }
            @Override
            public SqlStatementCustomizer createForMethod(Annotation annotation, @SuppressWarnings("rawtypes") Class sqlObjectType, Method method) {
                final OutParameter outParam = (OutParameter) annotation;
                return new SqlStatementCustomizer() {
                    @Override
                    public void apply(@SuppressWarnings("rawtypes") SQLStatement q) throws SQLException {
                        assert q instanceof Call;
                        ((Call) q).registerOutParameter(outParam.name(), outParam.sqlType());
                    }
                };
            }
            @Override
            public SqlStatementCustomizer createForParameter(Annotation annotation, @SuppressWarnings("rawtypes") Class sqlObjectType, Method method, final Object arg) {
                throw new UnsupportedOperationException("Not defined for parameter");
            }
        }
    }
}
```

**Comment by stevenschlansker** [comment]
Hi @jcshaw I have spent some time looking into this.  I am not sure how this code worked in the first place, from my reading, it is close but not quite right -- statement customizers are applied *after* parameter bindings are applied, so adding new parameter bindings in a statement customizer does not work.  Reading through how the `Call` class works, it seems a little overly complicated to me -- but maybe there's a reason it is the way it is today?

I am still poking at it, and may have something to show, just wanted to let you know that I think the `@OutParameter` annotation cannot work as-is.

**Comment by stevenschlansker** [comment]
Also, I think at least part of the problem here is that the code does not consistently use names -- notice how you call it `integerOut` in some places, `outParam1` in others...

**Comment by stevenschlansker** [comment]
I am curious, if you try using the Fluent `Call` API rather than the SqlObject API, does it work okay?
That might get you going with the current release while we work on implementing this feature.  I like the idea of this annotation but am not sure we have the developer bandwidth to commit to implementing it in v2, when we have a new major v3 release coming up that will make it much easier to implement.

**Comment by jcshaw** [comment]
@stevenschlansker You are correct, I have mixed up the variables in the in the example that I hastily put together. I have made them all outParam1 just in case and it still has a position of zero in OutParamArgument .

Thank you for taking a look at it, I appreciate your help.

**Comment by stevenschlansker** [comment]
Okay, I managed to make a bit more headway.  At least part of the problem is that the `Binding` class which takes care of the named and positional parameters keeps track of which ones are seen in the SQL statement, and for out parameters, they do not appear in the SQL text.  So it doesn't ever try to bind them.

**Comment by stevenschlansker** [comment]
Oh, but you didn't make that mistake -- just my test program did.  Hm, maybe that's not the problem then.

**Comment by stevenschlansker** [comment]
Okay @jcshaw I'm a little confused now.  I got it working in Postgres, it turns out that my syntax was wrong.  I don't use stored procedures much but if you want Postgres to give you out parameters apparently it wants a funky `{call myStoredProc()}` that I hadn't seen before.

The full code is here:

https://gist.github.com/stevenschlansker/42b410cbaa2b3015db3c9c0d1026ce41

I don't understand why it does not work for you.  I wonder if this is an Oracle specific thing?

**Comment by stevenschlansker** [comment]
Okay I've now assembled this as a pull request:
https://github.com/jdbi/jdbi/pull/406

I believe this shows that the annotation works fine for Postgres.  @jcshaw can you please try out this branch and use our copy of `@OutParameter` and capture exactly what the exception trace you get is?  I am worried this is going to turn out to be Oracle specific, which makes our lives hard, as we do not have any access to Oracle instances to test against...

**Comment by qualidafial** [comment]
Fixed in #406
